lib/src/models/collection_reference.dart

```diff
@@ -0,0 +1,179 @@
+import 'package:firedart/src/util/firestore_encoding.dart';
+
+class CollectionReference extends Reference {
+  StructuredQuery _structuredQuery;
+
+  CollectionReference(FirestoreGateway gateway, String path)
+      : super(gateway, path) {
+    _structuredQuery = StructuredQuery();
+    _structuredQuery.from
+        .add(StructuredQuery_CollectionSelector()..collectionId = id);
+    if (fullPath.split('/').length % 2 == 1) {
+      throw Exception('Path is not a collection: $path');
+    }
+  }
+
+  DocumentReference document(String id) {
+    return DocumentReference(gateway, '$path/$id');
+  }
+
+  /// Returns [CollectionReference] with additional filter on specified
+  /// [field]. [field] refers to a field in a document.
+  ///
+  /// The [field] may be a [String] consisting of a single field name
+  /// (referring to a top level field in the document),
+  /// or a series of field names separated by dots '.'
+  /// (referring to a nested field in the document).
+  ///
+  /// Only documents satisfying provided condition are included in the result
+  /// set.
+  CollectionReference where(
+    String fieldPath, {
+    dynamic isEqualTo,
+    dynamic isLessThan,
+    dynamic isLessThanOrEqualTo,
+    dynamic isGreaterThan,
+    dynamic isGreaterThanOrEqualTo,
+    dynamic arrayContains,
+    List<dynamic> arrayContainsAny,
+    List<dynamic> whereIn,
+    bool isNull,
+  }) {
+    StructuredQuery_CompositeFilter compositeFilter;
+    if (_structuredQuery.hasWhere() &&
+        _structuredQuery.where.hasCompositeFilter()) {
+      compositeFilter = _structuredQuery.where.compositeFilter;
+    } else {
+      compositeFilter = StructuredQuery_CompositeFilter()
+        ..op = StructuredQuery_CompositeFilter_Operator.AND;
+    }
+    final filter = StructuredQuery_Filter();
+    if (isNull != null) {
+      filter.unaryFilter = _createUnaryFilter(
+        fieldPath,
+        isNull,
+      );
+    } else {
+      filter.fieldFilter = _createFieldFilter(
+        fieldPath,
+        isEqualTo,
+        isLessThan,
+        isLessThanOrEqualTo,
+        isGreaterThan,
+        isGreaterThanOrEqualTo,
+        arrayContains,
+        arrayContainsAny,
+        whereIn,
+      );
+    }
+    compositeFilter.filters.add(filter);
+    _structuredQuery.where = StructuredQuery_Filter()
+      ..compositeFilter = compositeFilter;
+    return this;
+  }
+  /// Create [StructuredQuery_UnaryFilter] by provided [fieldPath] and [isNull] condition
+  StructuredQuery_UnaryFilter _createUnaryFilter(
+      String fieldPath, bool isNull) {
+    if (!isNull) {
+      throw Exception(
+          "'isNull': isNull can only be set to true. Use isEqualTo to filter on non-null values.");
+    }
+    final unaryFilter = StructuredQuery_UnaryFilter();
+    unaryFilter.op = StructuredQuery_UnaryFilter_Operator.IS_NULL;
+    return unaryFilter
+      ..field_2 = (StructuredQuery_FieldReference()..fieldPath = fieldPath);
```

Comment:
Maybe it would be better to avoid using nested `..` to improve readability?
```suggestion
    unaryFilter.op = StructuredQuery_UnaryFilter_Operator.IS_NULL;
    final secondField = StructuredQuery_FieldReference()..fieldPath = fieldPath;
    return unaryFilter
      ..field_2 = secondField;
```
or more granular:
```suggestion
    unaryFilter.op = StructuredQuery_UnaryFilter_Operator.IS_NULL;
    final secondField = StructuredQuery_FieldReference()..fieldPath = fieldPath;
    unaryFilter.field_2 = secondField;
    return unaryFilter;
```